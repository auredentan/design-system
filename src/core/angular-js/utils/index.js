import './lx-depth_service';
import './lx-enter-keypress';
import './lx-event-scheduler_service';
import './lx-utils_service';
