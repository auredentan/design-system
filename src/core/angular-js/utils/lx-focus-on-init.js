function lxFocusOnInitDirective() {
    function link(scope, el) {
        el.focus();
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module('lumx.utils.focus-on-init').directive('lxFocusOnInit', lxFocusOnInitDirective);

/////////////////////////////

export { lxFocusOnInitDirective };
