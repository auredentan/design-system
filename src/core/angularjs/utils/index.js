/* eslint-disable import/no-unassigned-import */

import './depth_service';
import './enter-keypress_directive';
import './event-scheduler_service';
import './focus-on-init_directive';
import './focus-trap_service';
import './utils_service';
