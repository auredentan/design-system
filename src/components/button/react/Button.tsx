import { Color, Colors, ComplexPropDefault, Size, Sizes, Theme, Themes } from 'LumX/components';

/////////////////////////////

import React, { Children } from 'react';

import classNames from 'classnames';

import isEmpty from 'lodash/isEmpty';
import isString from 'lodash/isString';

import { LxIcon, LxIconButton } from 'LumX';
import { CSS_PREFIX } from 'LumX/core/constants';
import {
    ChildTransformParameters,
    ChildValidateParameters,
    isElementOfType,
    isElementText,
    Omit,
    validateComponent,
} from 'LumX/core/react/utils';
import { handleBasicClasses } from 'LumX/core/utils';

import { LxButtonRoot, LxButtonRootProps } from './ButtonRoot';

/////////////////////////////

/**
 * The authorized values for the `emphasis` prop.
 */
enum Emphasises {
    low = 'low',
    medium = 'medium',
    high = 'high',
}
type Emphasis = Emphasises;

/**
 * The authorized values for the `variant` prop.
 */
enum Variants {
    button = 'button',
    icon = 'icon',
}
type Variant = Variants;

/////////////////////////////

/**
 * Defines the props of the <LxButton> component.
 */
interface IProps {
    /**
     * The button color.
     */
    color?: Color;

    /**
     * The emphasis of the button.
     */
    emphasis?: Emphasis;

    /**
     * The button size.
     */
    size?: Size;

    /**
     * The button theme.
     */
    theme?: Theme;

    /**
     * The button variant.
     */
    variant?: Variant;
}
type LxButtonProps = IProps & LxButtonRootProps;

/////////////////////////////

/**
 * Define the types of the default props.
 */
interface ILxButtonDefaultPropsType extends Partial<Omit<LxButtonProps, 'color'>> {
    color: ComplexPropDefault<Color>;
}

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * The default class name and classes prefix for this component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const CLASSNAME: string = `${CSS_PREFIX}-button`;

/**
 * The display name of the component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const COMPONENT_NAME: string = 'LxButton';

/**
 * The default value of props.
 *
 * @type {ILxButtonDefaultPropsType}
 * @constant
 * @readonly
 */
const DEFAULT_PROPS: ILxButtonDefaultPropsType = {
    color: {
        default: Colors.dark,
        [`emphasis-${Emphasises.high}`]: Colors.primary,
    },
    emphasis: Emphasises.high,
    size: Sizes.m,
    theme: Themes.light,
    variant: Variants.button,
};

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Transform the text children to <span>s when validating the children of the <LxButton> component.
 *
 * @param  {ChildTransformParameters} params The parameters received from the `validateComponent` function.
 * @return {React.ReactElement}       The transformed children (or the original one if there is no transformation to
 *                                    do).
 */
function _transformChild({ child }: ChildTransformParameters): React.ReactElement {
    if (isString(child)) {
        return <span>{child}</span>;
    }

    return child;
}

/**
 * Returns a closure for the function to validate the children of the <LxButton> component.
 * This closure will help remembering the types of the previous children. This list will help to determine if there is
 * no excess children of a given type.
 *
 * @param  {Array<string>} childrenTypes The list of types of the previously validated children.
 * @return {Function}      The closured function to validate the children of the <LxButton> component.
 */
function _validateChild(childrenTypes: string[]): (params: ChildValidateParameters) => void {
    /**
     * Validate the children of the <LxButton> component
     *
     * @param {ChildValidateParameters} params The parameters received from the `validateComponent` function.
     */
    return ({ child, childrenCount, index }: ChildValidateParameters): void => {
        const isChildText: boolean = isElementText(child) || isElementOfType(child, <span />);
        const isChildIcon: boolean = isElementOfType(child, LxIcon);

        const alreadyHasSomeText: boolean = index === 0 ? false : childrenTypes.some((type: string) => type === 'text');
        childrenTypes[index] = isChildText ? 'text' : 'icon';

        if (isChildText && alreadyHasSomeText) {
            throw new Error(
                `You cannot have more than 1 label in a 'button' \`variant\` of <${COMPONENT_NAME}> (got at least 2)!`,
            );
        }

        if (!isChildIcon) {
            return;
        }

        if (childrenCount === 1) {
            console.warn(
                `If you want to display an icon button, you should use the 'icon' \`variant\` of the <${COMPONENT_NAME}> instead of the 'button' \`variant\`\nYou should even consider using the <${
                    LxIconButton.displayName
                }> component instead.`,
            );
        }

        const isPreviousChildIcon: boolean = childrenTypes[index - 1] === 'icon';
        if (index > 0 && isPreviousChildIcon) {
            throw new Error(
                `You cannot have 2 following icons children in a 'button' \`variant\` of <${COMPONENT_NAME}>, there must be a label between them!`,
            );
        }
    };
}

/**
 * Validate the <LxButton> component props and children.
 * Also, sanitize, cleanup and format the children and return the processed ones.
 *
 * @param  {LxButtonProps}   props The children and props of the <LxButton> component.
 * @return {React.ReactNode} The processed children of the component.
 */
function _validate(props: LxButtonProps): React.ReactNode {
    const childrenTypes: string[] = [];

    return validateComponent(COMPONENT_NAME, {
        allowedTypes: props.variant === Variants.icon ? [LxIcon] : ['text', <span />, LxIcon],
        maxChildren: props.variant === Variants.icon ? 1 : 3,
        minChildren: 1,
        props,
        transformChild: props.variant === Variants.button ? _transformChild : undefined,
        validateChild: props.variant === Variants.button ? _validateChild(childrenTypes) : undefined,
    });
}

/////////////////////////////

/**
 * Displays a button.
 * If the `href` property is set, it will display a `<a>` HTML tag. If not, it will use a `<button>` HTML tag instead.
 *
 * @return {JSX.Element} The <LxButton> component.
 */
const LxButton: React.FC<LxButtonProps> = ({
    children,
    className = '',
    color,
    emphasis = DEFAULT_PROPS.emphasis,
    size = DEFAULT_PROPS.size,
    theme = DEFAULT_PROPS.theme,
    variant = DEFAULT_PROPS.variant,
    ...props
}: LxButtonProps): JSX.Element => {
    const newChildren = _validate({ children, color, emphasis, size, theme, variant, ...props });

    if (variant === Variants.button) {
        let index = -1;
        Children.forEach(
            newChildren,
            (child: any): void => {
                index++;

                if (isElementOfType(child, LxIcon)) {
                    className += isEmpty(className) ? '' : ' ';
                    className += `${CLASSNAME}--has-${index === 0 ? 'left' : 'right'}-icon`;
                }
            },
        );
    }

    if (isEmpty(color)) {
        color = DEFAULT_PROPS.color[`emphasis-${emphasis}`] || DEFAULT_PROPS.color.default;
    }

    return (
        <LxButtonRoot
            className={classNames(
                className,
                handleBasicClasses({
                    color,
                    emphasis,
                    prefix: CLASSNAME,
                    size,
                    theme: emphasis === Emphasises.high ? theme : undefined,
                    variant,
                }),
            )}
            {...props}
        >
            {newChildren}
        </LxButtonRoot>
    );
};
LxButton.displayName = COMPONENT_NAME;

/////////////////////////////

export {
    CLASSNAME,
    DEFAULT_PROPS,
    Color,
    Colors,
    Emphasis,
    Emphasises,
    LxButton,
    LxButtonProps,
    Size,
    Sizes,
    Theme,
    Themes,
    Variant,
    Variants,
};
