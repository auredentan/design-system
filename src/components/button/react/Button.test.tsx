import React from 'react';

import { mount, shallow } from 'enzyme';
import mockConsole from 'jest-mock-console';
import { build, fake, oneOf } from 'test-data-bot';

import get from 'lodash/get';
import without from 'lodash/without';

import { ICommonSetup, Wrapper, commonTestsSuite } from 'LumX/core/testing/utils.test';
import { getBasicClass } from 'LumX/core/utils';
import { mdiCheck, mdiChevronDown, mdiPlus } from 'LumX/icons';

import { Button, ButtonProps, CLASSNAME, DEFAULT_PROPS, Emphasises, Sizes, Themes, Variants } from './Button';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<ButtonProps>;

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;

    /**
     * The <ButtonRoot> element that is used as a wrapper for the label.
     */
    buttonRoot: Wrapper;

    /**
     * The <Icon> icon(s) (can have 0, 1 or 2).
     */
    icon: Wrapper;
}

/////////////////////////////
//                         //
//    Private attributes   //
//                         //
/////////////////////////////

/**
 * The default label to use for the tests.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const DEFAULT_LABEL: string = 'Label';

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Get the default value of the given prop of a <Button>, depending on the effective props of the component (some
 * default value depends on the value of another prop).
 *
 * @param {string}      prop  The name of the prop you want the default value of.
 * @param {ISetupProps} props The current props of the <Button>.
 */
function _getDefaultPropValue({ prop, props }: { prop: string; props?: ISetupProps }): string {
    return prop === 'color'
        ? DEFAULT_PROPS[prop][`emphasis-${get(props, 'emphasis', DEFAULT_PROPS.emphasis)}`] ||
              DEFAULT_PROPS[prop].default
        : DEFAULT_PROPS[prop];
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  {ISetupProps} props  The props to use to override the default props of the component.
 * @param  {boolean}     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return {ISetup}      An object with the props, the component wrapper and some shortcut to some element inside of the
 *                       component.
 */
const setup: (props?: ISetupProps, shallowRendering?: boolean) => ISetup = (
    { ...propsOverrides }: ISetupProps = {},
    shallowRendering: boolean = true,
): ISetup => {
    const props: ButtonProps = {
        children: DEFAULT_LABEL,
        ...propsOverrides,
    };

    const renderer: (el: React.ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<Button {...props} />);

    return {
        buttonRoot: wrapper.find('ButtonRoot'),

        icon: wrapper.find('Icon'),

        props,
        wrapper,
    };
};

describe(`<${Button.displayName}>`, (): void => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', (): void => {
        it('should render correctly a text label', (): void => {
            const { buttonRoot, icon, wrapper }: ISetup = setup();
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).not.toExist();

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it('should render correctly a <span> label', (): void => {
            const children: React.ReactNode = <span>{DEFAULT_LABEL}</span>;

            const { buttonRoot, icon, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).not.toExist();

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it('should render correctly a left icon and a text label', (): void => {
            const { buttonRoot, icon, wrapper }: ISetup = setup({ leftIcon: mdiPlus });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it('should render correctly a left icon and a <span> label', (): void => {
            const { buttonRoot, icon, wrapper }: ISetup = setup({
                children: <span>{DEFAULT_LABEL}</span>,
                leftIcon: mdiPlus,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it('should render correctly a text label and a right icon', (): void => {
            const { buttonRoot, icon, wrapper }: ISetup = setup({ rightIcon: mdiChevronDown });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it('should render correctly a <span> label and a right icon', (): void => {
            const { buttonRoot, icon, wrapper }: ISetup = setup({
                children: <span>{DEFAULT_LABEL}</span>,
                rightIcon: mdiChevronDown,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it('should render correctly two icons and a text label', (): void => {
            const { buttonRoot, icon, wrapper }: ISetup = setup({ leftIcon: mdiPlus, rightIcon: mdiChevronDown });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(2);

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it('should render correctly two icons and a <span> label', (): void => {
            const { buttonRoot, icon, wrapper }: ISetup = setup({
                children: <span>{DEFAULT_LABEL}</span>,
                leftIcon: mdiPlus,
                rightIcon: mdiChevronDown,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(2);

            expect(buttonRoot.contains(DEFAULT_LABEL)).toBeTrue();
        });

        it("should render correctly an icon button with the 'button' `variant`", (): void => {
            mockConsole('info');

            let { buttonRoot, icon, wrapper }: ISetup = setup({ children: null, leftIcon: mdiPlus });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            /////////////////////////////

            ({ buttonRoot, icon, wrapper } = setup({ children: null, rightIcon: mdiChevronDown }));
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);
        });

        it("should render correctly an icon button with the 'icon' `variant`", (): void => {
            let { buttonRoot, icon, wrapper }: ISetup = setup({
                children: null,
                leftIcon: mdiPlus,
                variant: Variants.icon,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            /////////////////////////////

            ({ buttonRoot, icon, wrapper } = setup({
                children: null,
                rightIcon: mdiChevronDown,
                variant: Variants.icon,
            }));
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();
            expect(buttonRoot).toHaveClassName(CLASSNAME);

            expect(icon).toExist();
            expect(icon.length).toEqual(1);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        it('should use default props', (): void => {
            const { buttonRoot }: ISetup = setup();

            Object.keys(DEFAULT_PROPS).forEach(
                (prop: string): void => {
                    expect(buttonRoot).toHaveClassName(
                        getBasicClass({ prefix: CLASSNAME, type: prop, value: _getDefaultPropValue({ prop }) }),
                    );
                },
            );
        });

        it('should use the given props', (): void => {
            const modifiedPropsBuilder: () => ISetupProps = build('props').fields({
                // tslint:disable-next-line: no-any
                color: fake((fakeData: any): string => fakeData.commerce.color()),
                emphasis: oneOf(...without(Object.values(Emphasises), DEFAULT_PROPS.emphasis)),
                leftIcon: oneOf(mdiPlus, mdiCheck),
                size: oneOf(...without(Object.values(Sizes), DEFAULT_PROPS.size)),
                theme: oneOf(...without(Object.values(Themes), DEFAULT_PROPS.theme)),
                variant: Variants.icon,
            });

            const modifiedProps: ISetupProps = modifiedPropsBuilder();

            let { buttonRoot }: ISetup = setup({ children: null, ...modifiedProps });

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    if (prop === 'leftIcon' || prop === 'rightIcon') {
                        return;
                    }

                    if (prop === 'theme') {
                        expect(buttonRoot).not.toHaveClassName(
                            getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] as string }),
                        );
                    } else {
                        expect(buttonRoot).toHaveClassName(
                            getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                        );
                    }
                },
            );

            /////////////////////////////

            delete modifiedProps.leftIcon;
            modifiedProps.rightIcon = mdiChevronDown;

            ({ buttonRoot } = setup({ children: null, ...modifiedProps }));

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    if (prop === 'leftIcon' || prop === 'rightIcon') {
                        return;
                    }

                    if (prop === 'theme') {
                        expect(buttonRoot).not.toHaveClassName(
                            getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] as string }),
                        );
                    } else {
                        expect(buttonRoot).toHaveClassName(
                            getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                        );
                    }
                },
            );

            /////////////////////////////

            modifiedProps.emphasis = Emphasises.high;
            modifiedProps.variant = Variants.button;

            ({ buttonRoot } = setup({ ...modifiedProps }));

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    if (prop === 'leftIcon' || prop === 'rightIcon') {
                        return;
                    }

                    expect(buttonRoot).toHaveClassName(
                        getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                    );
                },
            );

            /////////////////////////////

            delete modifiedProps.rightIcon;

            ({ buttonRoot } = setup({ ...modifiedProps }));

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    if (prop === 'leftIcon' || prop === 'rightIcon') {
                        return;
                    }

                    expect(buttonRoot).toHaveClassName(
                        getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                    );
                },
            );

            /////////////////////////////

            delete modifiedProps.leftIcon;
            modifiedProps.rightIcon = mdiChevronDown;

            ({ buttonRoot } = setup({ ...modifiedProps }));

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    if (prop === 'leftIcon' || prop === 'rightIcon') {
                        return;
                    }

                    expect(buttonRoot).toHaveClassName(
                        getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                    );
                },
            );
        });

        it("should not have any `theme` in 'low' or 'medium' `emphasis` but one in 'high'", (): void => {
            let { buttonRoot }: ISetup = setup({ emphasis: Emphasises.high });

            expect(buttonRoot.prop('className') as string).toContain(
                getBasicClass({ prefix: CLASSNAME, type: 'theme', value: '' }),
            );

            /////////////////////////////

            ({ buttonRoot } = setup({ emphasis: Emphasises.medium }));

            expect(buttonRoot.prop('className') as string).not.toContain(
                getBasicClass({ prefix: CLASSNAME, type: 'theme', value: '' }),
            );

            /////////////////////////////

            ({ buttonRoot } = setup({ emphasis: Emphasises.low }));

            expect(buttonRoot.prop('className') as string).not.toContain(
                getBasicClass({ prefix: CLASSNAME, type: 'theme', value: '' }),
            );
        });

        it('should forward any CSS class with icon(s)', (): void => {
            const modifiedProps: ISetupProps = {
                className: 'component component--is-tested',
            };

            let { buttonRoot }: ISetup = setup({ leftIcon: mdiPlus, ...modifiedProps });

            expect(buttonRoot).toHaveClassName(modifiedProps.className);

            /////////////////////////////

            ({ buttonRoot } = setup({ rightIcon: mdiChevronDown, ...modifiedProps }));

            expect(buttonRoot).toHaveClassName(modifiedProps.className);

            /////////////////////////////

            ({ buttonRoot } = setup({ leftIcon: mdiPlus, rightIcon: mdiChevronDown, ...modifiedProps }));

            expect(buttonRoot).toHaveClassName(modifiedProps.className);
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        beforeEach(
            (): void => {
                try {
                    /*
                     * If `console.warn` or `console.info` has been mocked at least one, this exists. So disable TS
                     * here.
                     */

                    // @ts-ignore
                    global.console.warn.mockRestore();
                    // @ts-ignore
                    global.console.info.mockRestore();
                } catch (exception) {
                    // Nothing to do here.
                }
            },
        );

        it("should fail when no child nor icons is given in the 'button' `variant`", (): void => {
            expect(
                (): void => {
                    setup({ children: null });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should fail when 2 icons are given without a label in the 'button' `variant`", (): void => {
            expect(
                (): void => {
                    setup({ children: null, leftIcon: mdiPlus, rightIcon: mdiChevronDown });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should fail when a label is given in the 'icon' `variant`", (): void => {
            expect(
                (): void => {
                    setup({ leftIcon: mdiPlus, variant: Variants.icon });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should fail when more than 1 icon is given in the 'icon' `variant`", (): void => {
            expect(
                (): void => {
                    setup({ children: null, leftIcon: mdiPlus, rightIcon: mdiChevronDown, variant: Variants.icon });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should inform the user when rendering an icon button with the 'button' `variant`", (): void => {
            global.console.info = jest.fn();

            setup({ children: null, leftIcon: mdiPlus });
            expect(global.console.info).toHaveBeenCalled();

            /////////////////////////////

            // @ts-ignore
            global.console.info.mockClear();

            setup({ children: null, rightIcon: mdiChevronDown });
            expect(global.console.info).toHaveBeenCalled();
        });

        it("should not inform the user when rendering an icon button with the 'icon' `variant`", (): void => {
            global.console.info = jest.fn();

            setup({ children: null, leftIcon: mdiPlus, variant: Variants.icon });
            expect(global.console.info).not.toHaveBeenCalled();

            /////////////////////////////

            // @ts-ignore
            global.console.info.mockClear();

            setup({ children: null, rightIcon: mdiChevronDown, variant: Variants.icon });
            expect(global.console.info).not.toHaveBeenCalled();
        });

        it("should have no `theme` in any other `emphasis` than 'high'", (): void => {
            let modifiedProps: ISetupProps = {
                emphasis: Emphasises.high,
            };

            let { buttonRoot }: ISetup = setup(modifiedProps);
            expect(buttonRoot).toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: 'theme', value: DEFAULT_PROPS.theme }),
            );

            /////////////////////////////

            modifiedProps = {
                emphasis: Emphasises.medium,
            };

            ({ buttonRoot } = setup(modifiedProps));
            expect(buttonRoot).not.toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: 'theme', value: DEFAULT_PROPS.theme }),
            );

            /////////////////////////////

            modifiedProps = {
                emphasis: Emphasises.low,
            };

            ({ buttonRoot } = setup(modifiedProps));
            expect(buttonRoot).not.toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: 'theme', value: DEFAULT_PROPS.theme }),
            );
        });

        it('should only have the "left-icon" CSS class when a left icon is passed in a \'button\' `variant`', (): void => {
            const { buttonRoot }: ISetup = setup({ leftIcon: mdiPlus });
            expect(buttonRoot).toHaveClassName(`${CLASSNAME}--has-left-icon`);
            expect(buttonRoot).not.toHaveClassName(`${CLASSNAME}--has-right-icon`);
        });

        it('should only have the "right-icon" CSS class when a right icon is passed in a \'button\' `variant`', (): void => {
            const { buttonRoot }: ISetup = setup({ rightIcon: mdiChevronDown });
            expect(buttonRoot).not.toHaveClassName(`${CLASSNAME}--has-left-icon`);
            expect(buttonRoot).toHaveClassName(`${CLASSNAME}--has-right-icon`);
        });

        it('should have both "left-icon" and "right-icon" CSS classes when both left and right icons are passed in a \'button\' `variant`', (): void => {
            const { buttonRoot }: ISetup = setup({ leftIcon: mdiPlus, rightIcon: mdiChevronDown });
            expect(buttonRoot).toHaveClassName(`${CLASSNAME}--has-left-icon`);
            expect(buttonRoot).toHaveClassName(`${CLASSNAME}--has-right-icon`);
        });
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, { className: 'buttonRoot', prop: 'buttonRoot' }, { className: CLASSNAME });
});
