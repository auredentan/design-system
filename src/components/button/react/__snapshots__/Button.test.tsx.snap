// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Conditions should fail when anything else than a <Icon> is given as child in the 'icon' \`variant\` 1`] = `"You can only have <Icon> children in <Button> (got 'Label')!"`;

exports[`<Button> Conditions should fail when anything else than a text, <span> or <Icon> is given as child in the 'button' \`variant\` 1`] = `"You can only have text, <span> or <Icon> children in <Button> (got <div>)!"`;

exports[`<Button> Conditions should fail when more than 1 child is given in the 'icon' \`variant\` 1`] = `"<Button> must have exactly 1 child (got 2)!"`;

exports[`<Button> Conditions should fail when more than 1 text or <span> child is given in the 'button' \`variant\` 1`] = `"You cannot have more than 1 label in a 'button' \`variant\` of <Button> (got at least 2)!"`;

exports[`<Button> Conditions should fail when more than 1 text or <span> child is given in the 'button' \`variant\` 2`] = `"You cannot have more than 1 label in a 'button' \`variant\` of <Button> (got at least 2)!"`;

exports[`<Button> Conditions should fail when more than 1 text or <span> child is given in the 'button' \`variant\` 3`] = `"You cannot have more than 1 label in a 'button' \`variant\` of <Button> (got at least 2)!"`;

exports[`<Button> Conditions should fail when more than 3 children are given in the 'button' \`variant\` 1`] = `"<Button> cannot have more than 3 children (got 4)!"`;

exports[`<Button> Conditions should fail when no child is given 1`] = `"<Button> must have at least 1 child (got 0)!"`;

exports[`<Button> Conditions should fail when there are 2 or more <Icon> in a row in the 'button' \`variant\` 1`] = `"You cannot have 2 following icons children in a 'button' \`variant\` of <Button>, there must be a label between them!"`;

exports[`<Button> Snapshots and structure should render correctly a <span> label 1`] = `
<ButtonRoot
  className="lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly a <span> label and an icon 1`] = `
<ButtonRoot
  className="lumx-button--has-right-icon lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".1"
  />
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly a text label 1`] = `
<ButtonRoot
  className="lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly a text label and an icon 1`] = `
<ButtonRoot
  className="lumx-button--has-right-icon lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".1"
  />
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly an icon and a <span> label 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly an icon and a text label 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly an icon button with the 'button' \`variant\` 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly an icon button with the 'icon' \`variant\` 1`] = `
<ButtonRoot
  className="lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-icon"
>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
  />
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly two icons and a <span> label 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon lumx-button--has-right-icon lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".2"
  />
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render correctly two icons and a text label 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon lumx-button--has-right-icon lumx-button lumx-button--color-primary lumx-button--emphasis-high lumx-button--size-m lumx-button--theme-light lumx-button--variant-button"
>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
  <Icon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".2"
  />
</ButtonRoot>
`;
